### PR TITLE
Comments in problem

### LINKED ISSUE #415: Comments in problems
Allow everyone to comment on problems

### PR DESCRIPTION
**Pull request by Katrin-kudryash**
## PR includes

-   [x] Feature

## Related issues

Resolve #415

### CODE REVIEW

**Review by tsumo — CHANGES_REQUESTED** [review]
Some translations are missing

**Comment by 9teen90nine** [comment]
A screenshot would be really helpfull

File: src/components/Problem/Problem.tsx [hunk 1]
```diff
@@ -20,11 +21,16 @@ import { ProblemStats } from '../ProblemStats/ProblemStats';
...
+import CommentCreateForm from '../CommentCreateForm/CommentCreateForm';
+import { useCommentCreateMutation, useCommentDeleteMutation } from '../../modules/commentHooks';
+import { CommentSchema, DeleteComment } from '../../modules/commentTypes';
+import { CommentView } from '../CommentView/CommentView';
 
 import { tr } from './Problem.i18n';
 
 type ProblemProps = {
     problem: ProblemWithRelationsAndProblemSection;
+    comment?: Comment;
```
Comment (tsumo) [inline src/components/Problem/Problem.tsx:33]: Why? All relevant comments are inside the problem object

File: src/modules/commentMethods.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+    const comment = await prisma.comment.create({
+        data: { ...restData, text, user: { connect: { id: userId } }, problem: { connect: { id: problemId } } },
+        include: { problem: { include: { comments: { include: { user: true } }, author: true } }, user: true },
+    });
+
+    if (comment.user !== comment.problem.author) {
+        return sendMail({
+            to: comment.problem.author.email,
+            subject: `Comment on proplem ${comment.problem.name}`,
```
Comment (tsumo) [inline src/modules/commentMethods.ts:22]: Typo "proplem"
No translation

File: src/modules/commentTypes.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+export type CreateComment = z.infer<typeof createCommentSchema>;
+
+export const editAndDeleteCommentSchema = commentSchema.extend({
+    id: z.string(),
+});
+
+export type EditAndDeleteComment = z.infer<typeof editAndDeleteCommentSchema>;
+
+export const deleteCommentSchema = z.object({
```
Comment (tsumo) [inline src/modules/commentTypes.ts:29]: Looks like this schema is already defined of line 23

File: src/utils/dateTime.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
+import { TLocale } from './getLang';
```
Comment (tsumo) [inline src/utils/dateTime.ts:1]: These helpers probably should be in bricks

File: src/components/CommentCreateForm/CommentCreateForm.i18n/ru.json [hunk 5]
```diff
@@ -0,0 +1,3 @@
+{
+    "Comment": ""
```
Comment (tsumo) [inline src/components/CommentCreateForm/CommentCreateForm.i18n/ru.json:2]: No translation

File: src/components/CommentForm/CommentForm.i18n/ru.json [hunk 6]
```diff
@@ -0,0 +1,4 @@
+{
+    "Leave a comment": "",
```
Comment (tsumo) [inline src/components/CommentForm/CommentForm.i18n/ru.json:2]: No translation

**Review by tsumo — CHANGES_REQUESTED** [review]

File: src/components/CardHeaderComment.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
+import { nullable, Link, Text } from '@taskany/bricks';
+import { gapXs, gray8 } from '@taskany/colors';
+import styled from 'styled-components';
+
+interface CardHeaderProps extends React.HTMLAttributes<HTMLDivElement> {
+    name?: string | null;
```
Comment (awinogradov) [inline src/components/CardHeaderComment.tsx:6]: does it make sense to render comment w/o author name?

File: src/components/CardHeaderComment.tsx [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+export const CardHeaderComment: React.FC<CardHeaderProps> = ({ name, timeAgo, href }) => {
+    return (
+        <StyledCardHeader>
+            {nullable(name, (n) => (
+                <Text size="xs" color={gray8} weight="bold">
+                    {n}
+                </Text>
+            ))}
+            {nullable(timeAgo, (node) => (
```
Comment (awinogradov) [inline src/components/CardHeaderComment.tsx:27]: timeAgo is required

File: src/components/CardHeaderComment.tsx [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+                <>
+                    {nullable(name, () => (
+                        <StyledSeparator>—</StyledSeparator>
+                    ))}
+                    {href ? (
```
Comment (awinogradov) [inline src/components/CardHeaderComment.tsx:32]: `nullable(href, (h) => <Link href={h} />, timeAgo)`

**Review by awinogradov — CHANGES_REQUESTED** [review]

**Comment by Katrin-kudryash** [comment]
https://github.com/taskany-inc/hire/assets/81121562/711efc29-d7d0-450c-9d7e-83634e0e1eef

**Comment by Katrin-kudryash** [comment]
Untitled 3

**Review by tsumo — APPROVED** [review]

**Comment by Katrin-kudryash** [comment]
https://github.com/taskany-inc/hire/assets/81121562/b3740658-e488-4a83-adb4-70d6816167ff

**Comment by Katrin-kudryash** [comment]
Untitled

**Comment by Katrin-kudryash** [comment]
Untitled

File: src/components/Problem/Problem.tsx [hunk 10]
```diff
@@ -84,6 +100,23 @@ export const Problem: FC<ProblemProps> = ({ problem }) => {
...
+
+            const result = await commentCreateMutation.mutateAsync({
+                text: value.text,
+                problemId: problem.id,
+                userId: session.user.id,
+            });
+
+            return result;
+        },
+        [commentCreateMutation, problem.id, session?.user.id],
```
Comment (9teen90nine) [inline src/components/Problem/Problem.tsx:117]: this should be fixed

**Review by 9teen90nine — CHANGES_REQUESTED** [review]

**Review by 9teen90nine — APPROVED** [review]

File: src/components/CardHeaderComment.tsx (hunk 7 above)
Comment (Katrin-kudryash) [reply src/components/CardHeaderComment.tsx]: No, it doesn't make sense.

File: src/components/CardHeaderComment.tsx (hunk 8 above)
Comment (Katrin-kudryash) [reply src/components/CardHeaderComment.tsx]: This code is irrelevant

**Pull request merged**